과제 제출합니다~

jinsu4755 opened the pull request:
익숙하지 않아서 그런가...정말 여러번 고치면서도 뭐가 맞는지 모르겠지만...

제출 후에도 객체지향 생활체조 보고 고칠 수 있을 만한곳 찾아서 어떻게 만들어야 좋을까 고민하겠습니다~

jinsu4755 committed:
Fix(step1_study): 스터디날 안올라간 파일? 마지막에 배우고 변경된 파일입니다.

jinsu4755 committed:
Feat(step1_study): 문자열 계산기 과제.

Feat(step1_study): 실패하는 코드 작성.

Feat(step1_study): 테스트 통과 가능한 코드 작성

일단 한번에 몰아서 작성함..함수로 분류가 필요해보임.
테스트 코드도 Exceoption이 뜨는지 확인할 필요가 있음.

Feat(step1_study):과제 조건에 맞도록 테스트 코드 수정 및 테스트 코드를 통과 할 수 있는 코드 작성

Feat(step1_study):과제 조건에 맞도록 테스트 코드 수정 및 테스트 코드를 통과 할 수 있는 코드 작성

Fix(step1_study): 리펙토링 완료...

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+
+import java.lang.NumberFormatException
+
+class StringCalculation(
+    private val calculateString: String?
+) {
```
계산하고자 하는 입력문자열을 생성자에 집어넣게되면 어떤 문제점들이 생길까?

jinsu4755 replied:
사실 해당 값을 호출되는 메서드에 인자로 넣으려했는데.. 생성자에 들어갈때? 문제점을 잘 모르겠어서 생각좀 해보겠습니다!

malibinYun replied:
```
class StringCalculation(
    private val calculateString: String?
)
```
이것과
```
class StringCalculation(
    calculateString: String?
)
```
이것의 차이점이 무엇일까?

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+        return calculateString.split(" ")
+    }
+
+    private fun validOperand(operand: String): Int {
+        return try {
+            operand.toInt()
+        } catch (errorMessage: NumberFormatException) {
+            throw IllegalArgumentException("$operand 은 피연산자로 사용될 수 없습니다.")
+        }
+    }
```
validate를 하는데 무언가를 반환한다? 뭔가 앞뒤가 안맞는거 같아. 검증로직은 검증만 하면 되는데 뭔가 리턴한다고 생각이 들지 않아. 
또, valid는 형용사이기 때문에 일을 하는 메서드에는 어울리지 않다고 생각해. check 내지는 validate처럼 동사를 사용하자.

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+    private val calculateString: String?
+) {
+
+    private fun validCalculateString(): List<String> {
+        if (calculateString.isNullOrBlank()) {
+            throw IllegalArgumentException("입력값이 Null 이거나 공백일 수 없습니다")
+        }
+        return calculateString.split(" ")
+    }
```
검증을 하는데, split 까지 하는 메서드네. 우선 하는 일이 두가지이고, validate를 하는데 split까지 해줄거라고는 메서드 명만 보고선 알 수 없을거라 생각해. 
또 String은 어떤 String인지 몰라서 더 구체적인 네이밍이 필요하다고봐.

jinsu4755 replied:
해당 부분은 validate하고 그럼 문자열을 나누는 함수로 분리를 해보도록 하겠습니다!

malibinYun commented on src/main/kotlin/KtlintTest.kt:
```diff
@@ -1,2 +1,18 @@
+import java.lang.IllegalArgumentException
+
 fun main() {
+
+    val num: Int = readLine()!!.toInt()
+
+    if (num !in 1..10) {
+        throw IllegalArgumentException("test")
+    }
+
+    require(num !in 1..10) { "test" } // IllegalArgumentException 이 뜬다.
+
+    // runtime exception은 사용하지 말자!. 최상위 이기 때문에 더 구체적인 exception을 사용하는 것이 좋다.
+
+    check(num !in 1..10) { "test" } // status 상태가 말이 안된때만
+
+    // null이 논리적으로 말이 안될때는 무조건 예외처리
 }
```
이런 프로덕션 코드와 상관없는 코드들은 삭제하는게 좋아.

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+            "-" -> minus(result, operand)
+            "*" -> multiplication(result, operand)
+            "/" -> divide(result, operand)
+            else -> throw IllegalArgumentException(
+                "$operator 를 연산자로 사용하실 수 없습니다. +, -, *, / 의 연산을 사용해주세요."
+            )
```
예외처리 굳굳 👍👍 또 뭐가 틀렸는지 알려주는 것도 👍👍 좋은 습관이야.

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    private fun operation(result: Int, operand: Int, operator: String): Int {
+        return when (operator) {
```
설계를 바꿔서 파라미터를 2개 이하로 줄여보는건 어떨까?

l2hyunwoo replied:
정수를 정수로 나눌땐 몫이 출력되니까 정확한 값이 출력 안될 것 같아 그 부분 생각해서 operand는 실수형으로 바꾸는 게 좋을 것 같아

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    private fun add(result: Int, operand: Int): Int {
+        return result + operand
+    }
+
+    private fun minus(result: Int, operand: Int): Int {
+        return result - operand
+    }
+
+    private fun multiplication(result: Int, operand: Int): Int {
+        return result * operand
+    }
+
+    private fun divide(result: Int, operand: Int): Int {
+        return result / operand
+    }
```
이 메서드들은 굳이 래핑하지 않아도 괜찮을거같아!

jinsu4755 replied:
아 맞아요 처음에는 래핑하지 않았었는데, 궁금한게 사실 그러면 함수안에서 if나 when을 쓰는거 자체가 그 함수가 하는일이 경우에 따라 2~다수 로 바뀌는건 아닌가요?? 사실 이부분이 항상 궁금해서 "함수가 하나의 일을 한다"가... if,when 구문을 사용하면 어떻게 되는것인지, 그럼 이경우는 사실 네이밍을 기똥차게 하는게 답일까요??

malibinYun replied:
함수가 하나의 일을 한다 라는 말은 사실 어려운말이긴 하지.
추상화 단계에 대해서 생각해보는게 좋을거같아.
물론 위 추상화 단계에서는 "더하기, 빼기, ..." 이지만,
한 단계 위 추상화 단계에서는 "계산하기" 이니까.

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private fun calculate(calculateList: List<String>): Int {
+        var result = validOperand(calculateList[0])
```
result 인데 첫 피연산자 값이 들어간다?
무슨 의도인지는 알겠지만, 네이밍이 적절하진 않아보여.

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+        return result / operand
+    }
+
+    private fun calculate(calculateList: List<String>): Int {
+        var result = validOperand(calculateList[0])
+        for (i in 1 until calculateList.size step 2) {
+            val operator: String = calculateList[i]
+            val operand: Int = validOperand(calculateList[i + 1])
+            result = operation(result, operand, operator)
+        }
+        return result
+    }
```
메서드 분리가 잘 되어있어서 계산하는 딱 한가지의 일만 하고있네!
그래서 코드도 짧고 의도가 명확히 보여. 👍

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,68 @@
+        return result
+    }
+
+    fun calculateResult(): Int {
+
+        val calculateStringList: List<String> = validCalculateString()
+
+        return calculate(calculateStringList)
+    }
```
이게 Lint에 안걸리다니 흠..

malibinYun replied:
딱 메서드명만 두고 봤을 때, 나는 절대 기능 구분을 못할거같아.
* calculateResult()
* calculate()

Result란 말은 불용어 인거같아. https://github.com/MalibinStudy/RacingCar/pull/11#discussion_r507021534

jinsu4755 replied:
저도 사실 많이 고민해봤는데 다른분들 리뷰하면서 어떤 이름을 써야할지 좀더 고민하도록 하겠습니다!!

malibinYun commented on src/test/kotlin/firstweekstringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,74 @@
+        val calculationString: String = "2 + 3 * 4 / 2"
+        val stringCalculation = StringCalculation(calculationString)
+
+        val resultCalculationString: Int = stringCalculation.calculateResult()
+
+        assertThat(resultCalculationString).isEqualTo(10)
```
```suggestion

        assertThat(stringCalculation.calculateResult()).isEqualTo(10)
```
이건 인라인이 더 나아보여

malibinYun commented on src/test/kotlin/firstweekstringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,74 @@
+        val calculationString: String? = null
+        val stringCalculation = StringCalculation(calculationString)
+
+        val thrownTestFunction: () -> Unit = {
+            stringCalculation.calculateResult()
+        }
+
+        assertThatThrownBy(thrownTestFunction)
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("입력값이 Null 이거나 공백일 수 없습니다")
```
```suggestion
        assertThatThrownBy { stringCalculation.calculateResult() }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("입력값이 Null 이거나 공백일 수 없습니다")
```
개인적으로 나는 이게 더 나아보여.

jinsu4755 replied:
아 이 방법을 생각하지 못했었네요 수정하도록 하겠습니다!

malibinYun commented on src/test/kotlin/firstweekstringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,74 @@
+            .hasMessageContaining("입력값이 Null 이거나 공백일 수 없습니다")
+    }
+
+    @Test
+    fun ariseStringCalculatorIllegalArgumentExceptionCaseBlank() {
+        val calculationString: String = "  "
+        val stringCalculation = StringCalculation(calculationString)
+
+        val thrownTestFunction: () -> Unit = {
+            stringCalculation.calculateResult()
+        }
+
+        assertThatThrownBy(thrownTestFunction)
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("입력값이 Null 이거나 공백일 수 없습니다")
+    }
+
+    @Test
+    fun ariseStringCalculatorIllegalArgumentExceptionCaseWrongOperator() {
+        val calculationString: String = "2 $ 3 * 4 / 2"
+        val stringCalculation = StringCalculation(calculationString)
+
+        val thrownTestFunction: () -> Unit = {
+            stringCalculation.calculateResult()
+        }
+
+        assertThatThrownBy(thrownTestFunction)
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("연산자로 사용하실 수 없습니다. +, -, *, / 의 연산을 사용해주세요.")
+    }
```
예외사항 테스트 👍

malibinYun reviewed (changes requested):
과제 열심히 하느라 고생했어~~ 😀

전반적으로 코드가 짧고 간결해서 잘 읽혔어. 그래도 이름이나 책임 분리같은걸 중점적으로 더 해봅시다.

한 클래스가 너무 많은 일들을 하고 있어서, 역할을 분리해보자. 다른 사람들 리뷰도 참고하면 좋을거같아. 똑같은 말을 여러명한테 하다보니까 잊어먹고 코멘트 안하는것도 있을거같네..

지금은 "2 3" 입력시에 인덱스 바운드 Exception이 나타나네 리팩터링 해봅시다.

chaeppy reviewed (commented):
제가 달려고한 코멘트들이 모두 혁님이 이미 말씀하신 것들이라... 코드는 모두 봤습니다 !!! 수고하셨습니다 !!! ^ㅁ^

jinsu4755 committed:
Fix(step1_study):1차 피드백 적용

그런데 아직 함수에 3개 인자 받는 부분은 수정하지 못함..

jinsu4755 committed:
Feat(step1_study2):2차 스터디 배운 내용을 적용해보았습니다.

jinsu4755 committed:
fix(step1_study2): 안올라간 변경사항 추가..

operator에심볼을 private안하면 및줄 그어져서 추가했습니다..
테스트 케이스 10에서 10.1로 변경했는데 안올라갔었습니다.

jinsu4755 commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,39 @@
+        return calculateString.split(" ")
+    }
+
+    private fun operation(result: Double, operand: Double, operatorSymbol: String): Double {
+        val operator = Operator.findByOperatorSymbol(operatorSymbol)
+        return operator.calculateStrategy(result, operand)
+    }
```
이부분 파라메터를 어떻게 줄일지 감이 안잡혀서 구조를 어떻게 고치면 좋을지 생각해보고있겠습니다..ㅠ

jinsu4755 committed:
Fix(step1_study2): Enum Test

malibinYun commented on src/main/kotlin/firstweekstringcalculator/Operator.kt:
```diff
@@ -0,0 +1,24 @@
+
+    companion object {
+        fun findByOperatorSymbol(operatorSymbol: String): Operator {
+            return values().find { it.hasOperatorSymbol(operatorSymbol) }
+                ?: throw IllegalArgumentException("$operatorSymbol 를 연산자로 사용하실 수 없습니다. +, -, *, / 의 연산을 사용해주세요.")
```
에러 상세 메시지 👍👍

malibinYun commented on src/test/kotlin/firstweekstringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,66 @@
+        val calculationString: String = "2 $ 3 * 4 / 2"
+        val stringCalculation = StringCalculation()
+
+        assertThatThrownBy { stringCalculation.calculate(calculationString) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("연산자로 사용하실 수 없습니다. +, -, *, / 의 연산을 사용해주세요.")
```
```suggestion
        assertThatThrownBy { stringCalculation.calculate(calculationString) }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessage("$ 를 연산자로 사용하실 수 없습니다. +, -, *, / 의 연산을 사용해주세요.")
```
hasMessage를 이용해서 더 정확한 테스트를 진행해보자.

malibinYun commented on src/main/kotlin/firstweekstringcalculator/ValidateStringCalculation.kt:
```diff
@@ -0,0 +1,17 @@
+
+class ValidateStringCalculation {
+
+    fun validateCalculateString(calculateString: String?): String {
+        if (calculateString.isNullOrBlank()) throw IllegalArgumentException("입력값이 Null 이거나 공백일 수 없습니다")
+        return calculateString
+    }
```
검증을 했는데 입력값이 반환된다? 나는 조금 어색해보여.
검증은 검증만 하는게 좋아!

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    fun calculate(calculateString: String?): Double {
+        val validString = validateCalculateString.validateCalculateString(calculateString)
```
validateCalculateString.validateCalculateString(calculateString).....
네이밍 고민을 좀더 해보는건 어떨까?

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,39 @@
+            val operator: String = calculateList[i]
+            val operand: Double = calculateList[i + 1].also {
+                validateCalculateString.validateOperand(it)
+            }.toDouble()
+            accumulator = operation(accumulator, operand, operator)
```
https://github.com/MalibinStudy/RacingCar/pull/14/files#r511604428
여기에서 해당 코드를 써도 충분할거같아.

malibinYun commented on src/main/kotlin/firstweekstringcalculator/StringCalculation.kt:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private fun calculateAccumulator(calculateList: List<String>): Double {
+        var accumulator = calculateList[0].also {
+            validateCalculateString.validateOperand(it)
+        }.toDouble()
```
이게 좀 먼가 아쉽다. 아래에도 똑같은 중복코드가 있기도 하구.
이럴땐 어쩌면 좋을까? 고민 해보고 나중에 물어보면 가르쳐줄게.

malibinYun reviewed (approved):
전반적으로 많이 나아지셨군요.
네이밍이랑 몇 가지 분리에서 아쉽긴하지만 연습하다보면 점차 더 나아질거야.
리팩터링 그만 하려면 머지해달라고 말해줘!

jinsu4755 committed:
Refact(step1): 2차 피드백 반영.

Fix(step1_study3): 입력 뷰 분리및 테스트 작성

Fix(step1_study3): 연산자 클래스 수정

계산 기호에게 계산의 동작을 추가함.
0으로 나눌 수 없다는 조건을 나누기 계산 전략에게 추가함.

Fix(step1_study3): 스트링 피연산자를 계산 가능하게 바꿀 수 있는 클래스 추가

Double로 변경할땐 소수점 2자리 이하는 버리도록함.

Fix(step1_study3): 기존에 클래스에서 입력 받는 부분을 뷰로 빼냄

Refact(step1_study3): 전체적인 과제 재설계

문자열 계산기는 정말 계산의 역할만.
연산자에게는 연산자와 연산 전략 역할.
등등 지금까지 배운 모든 내용 최대한 활용해봄...

malibinYun commented:
오 BigDecimal 사용 굳
전반적으로 깔끔하구만 중간중간 require 로 바꿔보는걸 습관 들여보구 네이밍도 항상 고민을 많이 해봐~~
고생햇습니다 다음미션 합시다
